src/perf/lib/ThroughputServer.cpp

```diff
@@ -0,0 +1,193 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    QUIC Perf Throughput Server Implementation.
+
+--*/
+
+#ifdef QUIC_CLOG
+#include "ThroughputServer.cpp.clog.h"
+#endif
+
+#ifndef _KERNEL_MODE
+#define QUIC_TEST_APIS 1
+#endif
+#include "msquichelper.h"
+#include "ThroughputServer.h"
+#include "ThroughputCommon.h"
+
+ThroughputServer::ThroughputServer() {
+    if (Session.IsValid()) {
+        Session.SetAutoCleanup();
+        Session.SetPeerUnidiStreamCount(THROUGHPUT_SERVER_PEER_UNI);
+        Session.SetDisconnectTimeout(THROUGHPUT_DEFAULT_DISCONNECT_TIMEOUT);
+        Session.SetIdleTimeout(THROUGHPUT_DEFAULT_IDLE_TIMEOUT);
+    }
+}
+
+QUIC_STATUS
+ThroughputServer::Init(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[]
+    ) {
+    uint16_t port = THROUGHPUT_DEFAULT_PORT;
+    TryGetValue(argc, argv, "port", &port);
+
+    const char* localAddress = nullptr;
+    if (!TryGetValue(argc, argv, "listen", &localAddress)) {
+        WriteOutput("Server mode must have -listen\n");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+    if (!ConvertArgToAddress(localAddress, port, &Address)) {
+        WriteOutput("Failed to decode IP address: '%s'!\nMust be *, a IPv4 or a IPv6 address.\n", localAddress);
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    TryGetValue(argc, argv, "connections", &NumberOfConnections);
+
+    QUIC_STATUS Status = SecurityConfig.Initialize(argc, argv, Registration);
+
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::Start(
+    _In_ QUIC_EVENT StopEvent
+    ) {
+    QUIC_STATUS Status =
+        Listener.Start(
+            &Address,
+            [](HQUIC Handle, void* Context, QUIC_LISTENER_EVENT* Event) -> QUIC_STATUS {
+                return ((ThroughputServer*)Context)->ListenerCallback(Handle, Event);
+            },
+            this);
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+    RefCount = CountHelper{StopEvent};
+    if (NumberOfConnections > 0) {
+        for (uint32_t i = 0; i < NumberOfConnections; i++) {
+            RefCount.AddItem();
+        }
+    } else {
+        //
+        // Add a single item so we can wait on the Count Helper
+        //
+        RefCount.AddItem();
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::Wait(
+    _In_ int Timeout
```

Comment:
Does this wait function really need to return a status?

Reply:
I want the API to return a status, because its possible there could be error cases in the not too distant future.